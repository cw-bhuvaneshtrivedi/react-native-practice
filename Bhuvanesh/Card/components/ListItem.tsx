import React, { MutableRefObject, useState } from "react";
import { View, Text, TouchableOpacity } from "react-native";
import globalStyles from "../Styles";
import { Ionicons } from "@expo/vector-icons";
import { FontAwesome } from "@expo/vector-icons";

type ListItemProps = {
  version: {
    name: string;
    maxPower: string;
    displacement: string;
    transmission: string;
    fuelType: string;
    priceOverview: {
      price: string;
      label: string;
    };
    emi: string;
  };
  setCount: (fn: (count: number) => number) => void;
};

const {
  flexRow,
  spaceBetween,
  listCard,
  buttonBackground,
  buttonText,
  specs,
  eMI,
  versionName,
  offers,
  price,
  overviewLabel,
  textList,
  alignEnd,
} = globalStyles;

type SpecsProps = {
  maxPower: string;
  displacement: string;
  transmission: string;
  fuelType: string;
};

type PriceProps = {
  priceOverview: {
    price: string;
    label: string;
  };
  emi: string;
};
export const Specs = ({
  maxPower,
  displacement,
  transmission,
  fuelType,
}: SpecsProps) => {
  return (
    <View style={[flexRow, spaceBetween, alignEnd]} testID="Specs">
      <Text style={specs}>{maxPower}</Text>
      <Text style={specs}>{displacement}</Text>
      <Text style={specs}>{transmission}</Text>
      <Text style={specs}>{fuelType}</Text>
    </View>
  );
};

export const Price = ({ priceOverview, emi }: PriceProps) => {
  return (
    <>
      <View style={[flexRow, spaceBetween, alignEnd]} testID="Price">
        <Text style={price}>{priceOverview.price}</Text>
        <Text style={offers}>Show price in my city</Text>
        <Text style={eMI} testID="versionList">
          EMI
        </Text>
      </View>
      <View style={[flexRow, spaceBetween, alignEnd]}>
        <Text style={overviewLabel}>{priceOverview.label}</Text>
        <Text style={{ color: "#484848", fontStyle: "italic" }}>{emi}</Text>
      </View>
    </>
  );
};

export const ListItem = ({ version, setCount }: ListItemProps) => {
  const [isShortList, setIsShortList] = useState(false);
  const [isCompare, setIsCompare] = useState(false);

  const {
    name,
    maxPower,
    displacement,
    transmission,
    fuelType,
    priceOverview,
    emi,
  } = version;

  const handleShortlistPress = () => {
    setCount((count: number) => count + (isShortList ? -1 : 1));
    setIsShortList((state) => !state);
  };

  const handleComparePress = () => {
    setIsCompare((state) => !state);
  };

  const Buttons = () => {
    const colorBeforeClick = "#6f6f6f";
    const colorAfterClick = "#00afa0";
    return (
      <View style={[buttonBackground, flexRow]}>
        <TouchableOpacity
          style={[
            flexRow,
            { width: 120, justifyContent: "center", alignItems: "center" },
          ]}
          onPress={handleShortlistPress}
          testID="ShortlistBtn"
        >
          <Ionicons
            name={isShortList ? "bookmark" : "bookmark-outline"}
            size={12}
            color={isShortList ? colorAfterClick : colorBeforeClick}
          />
          <Text style={[buttonText, isShortList ? { color: "#484848" } : {}]}>
            {isShortList ? "Shortlisted" : "Shortlist"}
          </Text>
        </TouchableOpacity>
        <Text style={{ fontSize: 30, color: "#e1e1e1" }}>|</Text>
        <TouchableOpacity
          style={[
            flexRow,
            { width: 200, justifyContent: "center", alignItems: "center" },
          ]}
          onPress={handleComparePress}
          testID="CompareBtn"
        >
          <FontAwesome
            name="balance-scale"
            size={12}
            color={isCompare ? colorAfterClick : colorBeforeClick}
          />
          <Text style={[buttonText, isCompare ? { color: "#484848" } : {}]}>
            {isCompare ? "Remove from Compare" : "Add to Compare"}
          </Text>
        </TouchableOpacity>
      </View>
    );
  };

  return (
    <View style={listCard}>
      <View style={textList}>
        <Text style={versionName}>{name}</Text>
        <Specs
          maxPower={maxPower}
          displacement={displacement}
          transmission={transmission}
          fuelType={fuelType}
        />
        <Price priceOverview={priceOverview} emi={emi} />
      </View>
      <Buttons />
    </View>
  );
};
