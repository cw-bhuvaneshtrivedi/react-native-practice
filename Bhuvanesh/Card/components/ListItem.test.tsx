import { render, fireEvent } from "@testing-library/react-native";
import { ListItem, Price, Specs } from "./ListItem";
import tmp from "../tmp.json";

describe("List Item", () => {
  const version = tmp.models[0].versions[0];

  it("Test Specs Alignment in row", () => {
    const { maxPower, displacement, transmission, fuelType } = version;
    const { getByTestId } = render(
      <Specs
        maxPower={maxPower}
        displacement={displacement}
        transmission={transmission}
        fuelType={fuelType}
      />
    );

    const element = getByTestId("Specs");
    expect(element.props.style[0].flexDirection).toEqual("row");
  });

  it("Test Price", () => {
    const { priceOverview, emi } = version;
    const { getByTestId } = render(
      <Price priceOverview={priceOverview} emi={emi} />
    );
    const element = getByTestId("Price");
    expect(element.type).toEqual("View");
    // expect(element).toMatchSnapshot();
  });

  it("Test ButtonClicks", () => {
    const setCount = jest.fn();
    setCount.mockImplementation();
    const { getByTestId, getByText } = render(
      <ListItem version={version} setCount={setCount} />
    );
    let element = getByTestId("CompareBtn");
    fireEvent(element, "click");
    expect(getByText("Remove from Compare")).not.toBeNull();
  });
});
