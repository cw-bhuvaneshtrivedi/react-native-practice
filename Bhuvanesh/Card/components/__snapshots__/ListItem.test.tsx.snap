// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Item Should match snapshot 1`] = `
<View
  style={
    {
      "borderColor": "#e1e1e1",
      "borderRadius": 3,
      "borderWidth": 1,
      "height": 170,
      "justifyContent": "space-between",
      "margin": 7,
      "width": 321,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "column",
        "height": 130,
        "justifyContent": "space-between",
        "padding": 12,
      }
    }
  >
    <Text
      style={
        {
          "color": "#0288d1",
          "fontSize": 16,
        }
      }
    >
      VXi S-CNG
    </Text>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "alignItems": "flex-end",
          },
        ]
      }
      testID="Specs"
    >
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 13,
            "fontWeight": "600",
          }
        }
      >
        56 bhp
      </Text>
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 13,
            "fontWeight": "600",
          }
        }
      >
        998 cc
      </Text>
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 13,
            "fontWeight": "600",
          }
        }
      >
        Manual
      </Text>
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 13,
            "fontWeight": "600",
          }
        }
      >
         CNG 
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "alignItems": "flex-end",
          },
        ]
      }
      testID="Price"
    >
      <Text
        style={
          {
            "color": "#484848",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Rs. 6,91,983
      </Text>
      <Text
        style={
          {
            "color": "#0288d1",
            "fontSize": 13,
          }
        }
      >
        Show price in my city
      </Text>
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 13,
            "fontStyle": "italic",
          }
        }
        testID="versionList"
      >
        EMI
      </Text>
    </View>
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "alignItems": "flex-end",
          },
        ]
      }
    >
      <Text
        style={
          {
            "color": "#aaaaaa",
            "fontSize": 12,
          }
        }
      >
        On-Road Price, Mumbai
      </Text>
      <Text
        style={
          {
            "color": "#484848",
            "fontStyle": "italic",
          }
        }
      >
        Rs. 12,497
      </Text>
    </View>
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#f9f9f9",
          "height": 40,
        },
        {
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
          "width": 120,
        }
      }
      testID="ShortlistBtn"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#6f6f6f",
              "fontSize": 12,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          [
            {
              "color": "#6f6f6f",
              "fontSize": 13,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        Shortlist
      </Text>
    </View>
    <Text
      style={
        {
          "color": "#e1e1e1",
          "fontSize": 30,
        }
      }
    >
      |
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "opacity": 1,
          "width": 200,
        }
      }
      testID="CompareBtn"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#6f6f6f",
              "fontSize": 12,
            },
            undefined,
            {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        
      </Text>
      <Text
        style={
          [
            {
              "color": "#6f6f6f",
              "fontSize": 13,
              "textAlign": "center",
            },
            {},
          ]
        }
      >
        Add to Compare
      </Text>
    </View>
  </View>
</View>
`;
